grails-app/services/org/pih/warehouse/data/MigrationService.groovy

```diff
@@ -359,6 +399,128 @@ class MigrationService {
+        TransactionType oldProductInventoryType = TransactionType.get(Constants.PRODUCT_INVENTORY_TRANSACTION_TYPE_ID)
+        List transactions = Transaction.findAllByInventoryAndTransactionType(location.inventory, oldProductInventoryType)?.sort { it.transactionDate}
+
+
+        Map<String, List<String>> results = [:]
+        recordMigrationResult(results, transactions, true)
+
+        if (performMigration && transactions) {
+            List<Product> products = transactions.transactionEntries?.flatten()?.collect { TransactionEntry te -> te.inventoryItem.product }?.unique()
+            // Always create a stock snapshot before modifying any transactions in order to prevent quantity on hand
+            // differences for edge cases that were not addressed
+            log.debug("Creating inventory baseline for current stock for products found in old transactions")
+            Transaction currentInventoryBaseline = productInventoryTransactionMigrationService.createInventoryBaselineTransaction(
+                    location,
+                    null,
+                    products,
+                    null,
+                    "Inventory baseline created during old product inventory transactions migration",
+                    null,
+                    true,
+                    true
+            )
+
+            transactions.each { Transaction it ->
+                log.debug "Migrating transaction ${it.transactionNumber} with ${it.transactionEntries.size()} transaction entries"
+                List<TransactionEntry> entries = it.transactionEntries
+                // 2. Find all products in entries (and create inventory baseline for these)
+                List<Product> currentTransactionProducts = entries?.collect { TransactionEntry te -> te.inventoryItem.product }?.unique()
+
+                // 3. Create inventory baseline for old transaction that is being migrated (if enabled)
+                Map<String, AvailableItem> availableItems = productAvailabilityService.getAvailableItemsAtDateAsMap(
+                        location, currentTransactionProducts, it.transactionDate)
+
+                Transaction baselineTransaction = productInventoryTransactionMigrationService.createInventoryBaselineTransactionForGivenStock(
+                        location,
+                        null,
+                        availableItems.values(),
+                        it.transactionDate,
+                        "Migrated from single Product Inventory transaction to Baseline + Adjustment pair",
+                        null,
+                        // don't validate transaction date, there is old transaction at the same time, that will be removed
+                        false,
+                        true
+                )
+                if (baselineTransaction) {
+                    baselineTransaction.disableRefresh = true
+                    // Ugly workaround to omit auto-timestamping dateCreated
```

Comment:
@awalkowiak move it up and change to fixme, and link gh issue related / documentation

Reply:
and move it to a method